### ISSUE TITLE
Neural Networks

### ISSUE THREAD

**Issue by frjnn** [neuronika/neuronika#17]
## The trait Model
I propose to create a public trait that users can implement to create their custom models. I provide a sketch hereafter.
```rust
pub trait Model {
    fn forward(&self, (...) );
}

impl Model for MyNet {
    // Something something machine learning.
}
```

## ~~The callable question~~

~~Should we make use of [this](https://doc.rust-lang.org/std/ops/trait.Fn.html) very cool but nightly feature? This would allow for something on the lines of this:~~

```rust
let net = MyModel::new();
let prediction = net(data); // Maybe sometime in the future =(
```

~~I personally would **love** to have such an API.~~ **Let's stick with stable.**

## Layers' initialisation functions
These are a must, we should take a quick look at [this](https://pytorch.org/docs/stable/nn.init.html) and then decide. Also we should look at [that](https://docs.rs/ndarray/0.15.1/ndarray/struct.ArrayBase.html#method.uninit) and [this](https://doc.rust-lang.org/std/default/trait.Default.html), I think there features will probably be useful for our intents.

The essentials in my opinion:

- [x] Uniform
- [x] Normal
- [x] Constant
- [x] Ones
- [x] Zeros
- [x] Eye
- [x] ~~Dirac~~ we should probably consider this at a later time since it is used only on n-dimensional tensors with n >=3 
- [x] Xavier Uniform
- [x] Xavier Normal

These should be easy and straightforward to implement.

## Type of Layers
The essentials in my opinion:
- [x]  Linear
- [x] Recurrent
- [x] Conv1d
- [x] Conv2d  
- [x] Conv3d 
- [x] GRU
- [x] LSTM
- [x] Dropout

## Useful links

- Pytorch [layers](https://pytorch.org/docs/stable/nn.html)
- ndarray's windows [method](https://docs.rs/ndarray/0.15.1/ndarray/struct.ArrayBase.html#method.windows)
@ste-pac .
